README
Text::FakeData version 0.05
==========================

NAME
    Text::FakeData - Perl extension for generating fake data

SYNOPSIS
      use Text::FakeData;

      my $faker = Text::FakeData->new();

      print "Name:    ".$faker->name."\n";
      print "Company: ".$faker->company."\n";
      print "Address: ".$faker->street_address."\n";
      print "         ".$faker->city.", ".$faker->state." ".$faker->zip."\n";

DESCRIPTION
    This module creates fake (but reasonable) data that can be used for
    things such as filling databases with fake information during
    development of database related applications.

OBJECT METHODS
    new()
        Object constructor. As a shortcut, you can pass names of plugin
        modules to load to new(), although this does not actually restrict
        the functions available to the object, it just causes those plugins
        to be loaded if they haven't been loaded already. All Text::FakeData
        objects in one interpreter share the plugin data, so that multiple
        objects don't multiply the memory requirements.

    methods();
        Return a list of the methods that have been provided by all of the
        loaded plugins.

    register_plugin();
        Plugin modules call register_plugin() to provide data methods. See
        any of the included plugin modules for examples.

LOADING PLUGINS
    You can specify which plugins to load by including just the base part of
    their name as an argument when loading the module with 'use'. For
    example if you only wanted to use data from the Text::FakeData::Name
    module, you would load Text::FakeData like this:

      use Text::FakeData qw(Name);

    By default any modules matching Text::FakeData::* in any directory in
    @INC will be loaded. You can also pass plugin names when calling the
    new() method, and they will be loaded if not already in memory. See
    new().

WRITING PLUGINS
    Writing a plugin to provide new kinds of data is easy, all you have to
    do is create a module named Text::FakeData::SomeModuleName that inherits
    from Text::FakeData.

    To provide data, the plugin merely needs to call the register_plugin
    function with one or more pairs of function name and function data, like
    this:

      #!/usr/bin/perl -w
      use strict;
      use warnings;
      use Text::FakeData;

      my $faker = Text::FakeData->new();
      print "My fake data is ".$faker->some_data_function."\n";

      package Text::FakeData::SomeData;
      use base 'Text::FakeData';

      __PACKAGE__->register_plugin(
        some_data_function => [qw(foo bar baz gazonk)],
        another_data_item => sub { return '$some_data_function' },
      );

    The first argument is the method that will be made available to your
    object, the second is a data source. If the data source is not a
    reference, it will simply be returned as the data, if it is a reference
    to an array, a random element from the array will be returned, and if it
    is a subroutine reference, the subroutine will be run and the results
    will be returned. The data that your data source provides is checked for
    two things, tokens (that look like perl variables, starting with a $),
    and numeric indicators (#). Any tokens found will be replaced with their
    values, and any numeric indicators will be replaced with random numbers.
    You can include a literal $ or # by prefacing it with a backslash. If
    you load more than one module that defines the same function, it has an
    additive effect, when the function is called one of the data sources
    provided will be selected at random and then it will be called to get a
    piece of data.

    Some data source examples:

      __PACKAGE__->register_plugin(
        age              => ['#','##'],
        monetary_amount  => ['\$####.##','\$###.##', '\$##.##', '\$#.##'],
        adult_age        => sub { int(rand(70)+18) },
      );

    If your data source is a code reference, it will receive the calling
    object as an argument so you can build data out of other data if you
    need to. See Text::FakeData::DateTime for some examples of this.

BUGS AND KNOWN ISSUES
    There is no way to selectively remove data sources from a plugin that
    was loaded, even if you didn't load it.

SEE ALSO
    Text::Lorem

AUTHOR
    Jason Kohles, <email@jasonkohles.com>

COPYRIGHT AND LICENSE
    Copyright 2004-2005 by Jason Kohles

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Text::FakeData.

0.06  Wed Jul 13 15:19:00 2005
	- Fixed some packaging problems

0.05  Wed Jul 13 15:10:00 2005
	- Initial public release


lib/Text/FakeData/Company.pm
package Text::FakeData::Company;
our $VERSION = '0.06';
use base 'Text::FakeData';
require Text::FakeData::Name;

=head1 NAME

Text::FakeData::Company - Text::FakeData plugin

=head1 SYNOPSIS AND USAGE

See L<Text::FakeData>

=head1 DATA PROVIDERS

=over 4

=item company

Return a fake company name.

=cut

__PACKAGE__->register_plugin('company' => [
	'$last_name $company_suffix',
	'$last_name-$last_name',
	'$last_name, $last_name and $last_name',
]);

=item company_suffix

Return a company suffix (Inc, LLC, etc).

=cut

__PACKAGE__->register_plugin('company_suffix' => [
	'Inc', 'and Sons', 'LLC', 'Group'
]);

=back

=head1 SEE ALSO

L<Text::FakeData>

=head1 AUTHOR

Jason Kohles, E<lt>email@jasonkohles.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2004-2005 by Jason Kohles

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

1;


lib/Text/FakeData/DateTime.pm
package Text::FakeData::DateTime;
our $VERSION = '0.06';
use base 'Text::FakeData';
use Date::Calc;
require POSIX;

=head1 NAME

Text::FakeData::DateTime - Text::FakeData plugin

=head1 SYNOPSIS AND USAGE

See L<Text::FakeData>

=head1 DATA PROVIDERS

=over 4

=item unixtime

Return a unix time (seconds since the epoch) for a random time between the
epoch and now.

=cut

__PACKAGE__->register_plugin('unixtime' => sub { int(rand(time())) });

=item date

Return a random date as a string, using a random date format (see date_format).

=cut

__PACKAGE__->register_plugin(
	'date' => sub { timestr(shift()->date_format) },
);

=item time

Return a random time as a string, using a random time format (see time_format).

=cut

__PACKAGE__->register_plugin(
	'time' => sub { timestr(shift()->time_format) },
);

=item rfc822

Return an RFC 822 formatted random date.  This method may not work on systems
using a non-GNU strftime implementation (kindly let me know if that is the
case.)

=cut

__PACKAGE__->register_plugin(
	'rfc822' => sub { timestr('%a,  %d  %b  %Y  %H:%M:%S  %z') },
);

=item ampm

Returns am or pm randomly (in the current locale) using one of the formats
specified in ampm_format.

=cut

__PACKAGE__->register_plugin(
	'ampm' => sub { timestr(shift()->ampm_format) },
);

=item time_format

Return a random time format.

=cut

__PACKAGE__->register_plugin(
	'time_format' => [qw(%R %r %T)],
);

=item date_format

Return a random date format.

=cut

__PACKAGE__->register_plugin(
	'date_format' => [qw(%D %F)],
);

=item ampm_format

Return a random am/pm format.

=cut

__PACKAGE__->register_plugin(
	'ampm_format' => [qw(%p %P)],
);

=item datetime_format

Return a random date and time format.

=cut

__PACKAGE__->register_plugin(
	'datetime_format' => ['%c','%+','%FT%H','%FT%I','%F %H','%F %I'],
);

=item month

Return a random month name, unabbreviated, in the current locale.

=cut

__PACKAGE__->register_plugin(
	'month' => sub { timestr('%B') },
);

=item month_abbr

Return a random month name, abbreviated, in the current locale.

=cut

__PACKAGE__->register_plugin(
	'month_abbr' => sub { timestr('%b') },
);

=item weekday

Return a random weekday name, unabbreviated, in the current locale.

=cut

__PACKAGE__->register_plugin(
	'weekday' => sub { timestr('%A') },
);

=item weekday_abbr

Return a random weekday name, abbreviated, in the current locale.

=cut

__PACKAGE__->register_plugin(
	'weekday_abbr' => sub { timestr('%a') },
);

=item sqldate

Return a random date in the ISO8601 format commonly used by SQL servers
(YYYY-MM-DD).

=cut

__PACKAGE__->register_plugin(
	'sqldate' => sub { timestr('%F') },
);

=item datetime_locale

Return a datetime string in the preferred date representation for the
current locale, for a random date.

=cut

__PACKAGE__->register_plugin(
	'datetime_locale' => sub { timestr('%c') },
);

=item date_locale

Return a date string in the preferred date representation for the
current locale, for a random date.

=cut

__PACKAGE__->register_plugin(
	'date_locale' => sub { timestr('%x') },
);

=item time_locale

Return a time string in the preferred date representation for the
current locale, for a random date.

=cut

__PACKAGE__->register_plugin(
	'time_locale' => sub { timestr('%X') },
);

=item century

Return a random century number.

=cut

__PACKAGE__->register_plugin(
	'century' => sub { timestr('%C') },
);

=item dayofmonth

Return a random day of the month.

=cut

__PACKAGE__->register_plugin(
	'dayofmonth' => sub { timestr('%d') },
);

=back

=head1 UTILITY METHODS

=over 4

=item Text::FakeData::DateTime::timestr($format);

Given a strftime format specifier, this method passes it through to
L<POSIX::strftime> along with a random date to display in that format.

Perl passes this through to the strftime function of your system library, so
it is possible that some of the formatting tokens used here will not work on
your system.

=cut

sub timestr {
	my $format = shift;
	if(ref($format)) { $format = shift }
	POSIX::strftime($format,localtime(__PACKAGE__->unixtime));
}

=back

=head1 NOTES AND CAVEATS

=over 4

=item Be careful build timestamps from pieces

Be very careful about building date/time representations in formats that
are not already listed here.  For example if you wanted to get a date that
consists of just the month and day, you should NOT do this:

  my $faker = Text::FakeData->new();
  print join(' ',$faker->month,$faker->dayofmonth)."\n";

This is bad because you might end up with 'February 31' for example.  Instead
you should use the timestr utility function to provide you a formatted time
for a valid date, or better still, write a plugin function that does it:

  my $faker = Text::FakeData->new();
  print $faker->my_short_date()."\n";

  package Text::FakeData::MyExtras;
  use base qw(Text::FakeData);
  use Text::FakeData::DateTime;
  __PACKAGE__->register_plugin(
    my_short_date => sub { Text::FakeData::DateTime::timestr('%M %e') },
  );

=item POSIX::strftime

See the documentation above regarding the timestr utility method for some
caveats related to strftime and your system library.

=back

=head1 SEE ALSO

L<Text::FakeData>

=head1 AUTHOR

Jason Kohles, E<lt>email@jasonkohles.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2004-2005 by Jason Kohles

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

1;


lib/Text/FakeData/Internet.pm
package Text::FakeData::Internet;
our $VERSION = '0.06';
use base 'Text::FakeData';
use Text::FakeData::Company;
use Text::FakeData::Name;

=head1 NAME

Text::FakeData::Internet - Text::FakeData plugin

=head1 SYNOPSIS AND USAGE

See L<Text::FakeData>

=head1 DATA PROVIDERS

=over 4

=item email

Return a fake email address.

=cut

__PACKAGE__->register_plugin(
	'email' => [qw($username@$domain_name)],
);

=item username

Return a fake username.

=cut

__PACKAGE__->register_plugin(
	'username'				=> sub {
		my $self = shift;
		my $name = lc($self->first_name);
		$name =~ s/\W//g;
		return $name;
	},
);

=item hostname

Return a fake computer hostname.

=cut

__PACKAGE__->register_plugin(
	'email' => [qw($domain_word.$domain_name)],
);

=item server_name

Return a fake server name (some service name such as mail, dns, etc, prepended
to a fake domain name.)

=cut

__PACKAGE__->register_plugin(
	'server_name' => [
		'$network_service.$domain_name',
		'$network_service-###.$domain_name',
	],
);

=item network_service

Return a random network service name.  Only fairly common services are included.

=cut

__PACKAGE__->register_plugin(
	network_service => [qw(
		ftp ssh telnet shell smtp mail time ns dns tacacs bootp dhcp www
		kerberos pop pop2 pop3 imap nfs ntp imap imap2 snmp irc imap3 https
		snpp isakmp ipp printer fileserver logs log loghost syslog news
		nntp ldap ldaps socks vpn sql db radius cvs svn xmpp x11 backup
	)],
);

=item domain_name

Return a fake domain_name.

=cut

__PACKAGE__->register_plugin(
	'domain_name'			=> [qw($domain_word.$domain_suffix)],
);

=item domain_word

Return a random word that meets the requirements for being part of a domain
name.

=cut

__PACKAGE__->register_plugin(
	'domain_word'			=> sub {
		my $self = shift;
		my $company = lc($self->company);
		$company =~ s/'//g;
		$company =~ s/\W+/-/g;
		
		return $company;
	},
);

=item domain_suffix

Return a random domain suffix (.com, .net, .co.uk. etc)

=cut

__PACKAGE__->register_plugin(
	'domain_suffix'		=> [qw(
		ac ac.uk ad ae af ag ai al am an ao aq ar as at au aw az ba bb bd be
		bf bg bh bi bj bm bn bo br bs bt bv bw by bz ca cc cd cf cg ch ci ck
		cl cm cn co co.uk com cr cs cu cv cx cy cz de dj dk dm do dz ec edu ee
		eg eh er es et fi fj fk fm fo fr ga gd ge gf gg gh gi gl gm gn gov gp
		gq gr gs gt gu gw gy hk hm hn hr ht hu id ie il im in int io iq ir is
		it je jm jo jp ke kg kh ki km kn kp kr kw ky kz la lb lc li lk lr ls
		lt lu lv ly ma mc md mg mh mil mk ml mm mn mo mp mq mr ms mt mu mv mw
		mx my mz na nc ne net nf ng ni nl no np nr nt nu nz om org pa pe pf pg
		ph pk pl pm pn pr ps pt pw py qa re ro ru rw sa sb sc sd se sg sh si sj
		sk sl sm sn so sr sv st sy sz tc td tf tg th tj tk tm tn to tp tr tt tv
		tw tz ua ug uk um us uy uz va vc ve vg vi vn vu wf ws ye yt yu za zm zw

		aero biz coop info museum name pro

		al.us ak.us az.us ar.us ca.us co.us ct.us de.us dc.us fl.us ga.us hi.us
		id.us il.us in.us ia.us ks.us ky.us la.us me.us md.us ma.us mi.us mn.us
		ms.us mo.us mt.us ne.us nv.us nh.us nj.us nm.us ny.us nc.us nd.us oh.us
		ok.us or.us pa.us ri.us sc.us sd.us tn.us tx.us ut.us vt.us va.us wa.us
		wv.us wi.us wy.us
	)],
);

=item ip_address

Return a random IP Address.

=cut

__PACKAGE__->register_plugin(
	'ip_address'		=> sub {
		my @n = (1 .. 254);
		return join('.',@n[rand(@n),rand(@n),rand(@n),rand(@n)]);
	},
);

=back

=head1 SEE ALSO

L<Text::FakeData>

=head1 AUTHOR

Jason Kohles, E<lt>email@jasonkohles.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2004-2005 by Jason Kohles

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

1;


lib/Text/FakeData/PhoneNumber.pm
package Text::FakeData::PhoneNumber;
our $VERSION = '0.06';
use base 'Text::FakeData';

=head1 NAME

Text::FakeData::PhoneNumber - Text::FakeData plugin

=head1 SYNOPSIS AND USAGE

See L<Text::FakeData>

=head1 DATA PROVIDERS

=over 4

=item phone_number

Return a fake phone number.

=cut

__PACKAGE__->register_plugin(
	phone_number => [
		'###-###-####',
		'(###)###-####',
		'1-###-###-####',
		'###.###.####',
		'###-###-####',
		'(###)###-####',
		'1-###-###-####',
		'###.###.####',
		'###-###-#### x###',
		'(###)###-#### x###',
		'1-###-###-#### x###',
		'###.###.#### x###',
		'###-###-#### x####',
		'(###)###-#### x####',
		'1-###-###-#### x####',
		'###.###.#### x####',
		'###-###-#### x#####',
		'(###)###-#### x#####',
		'1-###-###-#### x#####',
		'###.###.#### x#####',
	],
);

=back

=head1 SEE ALSO

L<Text::FakeData>

=head1 AUTHOR

Jason Kohles, E<lt>email@jasonkohles.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2004-2005 by Jason Kohles

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

1;


lib/Text/FakeData/StreetAddress.pm
package Text::FakeData::StreetAddress;
our $VERSION = '0.06';
use base 'Text::FakeData';
use Text::FakeData::Name;

=head1 NAME

Text::FakeData::StreetAddress - Text::FakeData plugin

=head1 SYNOPSIS AND USAGE

See L<Text::FakeData>

=head1 DATA PROVIDERS

=over 4

=item us_zip_code

Return a random zip or zip+4 zip code in the US zip code format.  Note that
this is not necessarily a valid zip code, just a 5 or 9 digit number in the
correct format.

=cut

__PACKAGE__->register_plugin(
	us_zip_code => ['#####','#####-####'],
);

=item us_state

Return a random US state name.

=cut


__PACKAGE__->register_plugin(
	us_state => [
		qw{
			Alabama Alaska Arizona Arkansas California Colorado Connecticut
			Delaware Florida Georgia Hawaii Idaho Illinois Indiana Iowa Kansas
			Kentucky Louisiana Maine Maryland Massachusetts Michigan Minnesota
			Mississippi Missouri Montana Nebraska Nevada Ohio Oklahoma Oregon
			Pennsylvania Tennessee Texas Utah Vermont Virginia Wisconsin
			Wyoming Washington
		},
		'New Hampshire', 'New Jersey', 'New Mexico', 'New York',
		'North Carolina', 'North Dakota', 'Rhode Island', 'South Carolina',
		'South Dakota', 'West Virginia',
	],
);

=item us_state_abbr

Return a random US state abbreviation. (Includes US Territories and AE, AA,
AP military designations.)

From the USPS list at http://www.usps.com/ncsc/lookups/usps_abbreviations.html
=cut

__PACKAGE__->register_plugin(
	us_state_abbr => [qw(
		AL AK AS AZ AR CA CO CT DE DC FM FL GA GU HI ID IL IN IA KS KY
		LA ME MH MD MA MI MN MS MO MT NE NV NH NJ NM NY NC ND MP OH OK 
		OR PW PA PR RI SC SD TN TX UT VT VI VA WA WV WI WY AE AA AP
	)],
);

=item street_suffix

Return a random street suffix (Drive, Street, Road, etc.)

From the USPS list at http://www.usps.com/ncsc/lookups/usps_abbreviations.html

=cut

__PACKAGE__->register_plugin(
	street_suffix => [qw(
		Alley Avenue Branch Bridge Brook Brooks Burg Burgs Bypass Camp Canyon 
		Cape Causeway Center Centers Circle Circles Cliff Cliffs Club Common 
		Corner Corners Course Court Courts Cove Coves Creek Crescent Crest 
		Crossing Crossroad Curve Dale Dam Divide Drive Drive Drives Estate
		Estates Expressway Extension Extensions Fall Falls Ferry Field Fields
		Flat Flats Ford Fords Forest Forge Forges Fork Forks Fort Freeway
		Garden Gardens Gateway Glen Glens Green Greens Grove Groves Harbor
		Harbors Haven Heights Highway Hill Hills Hollow Inlet Inlet Island
		Island Islands Islands Isle Isle Junction Junctions Key Keys Knoll
		Knolls Lake Lakes Land Landing Lane Light Lights Loaf Lock Locks Locks 
		Lodge Lodge Loop Mall Manor Manors Meadow Meadows Mews Mill Mills
		Mission Mission Motorway Mount Mountain Mountain Mountains Mountains 
		Neck Orchard Oval Overpass Park Parks Parkway Parkways Pass Passage
		Path Pike Pine Pines Place Plain Plains Plains Plaza Plaza Point Points
		Port Port Ports Ports Prairie Prairie Radial Ramp Ranch Rapid Rapids
		Rest Ridge Ridges River Road Road Roads Roads Route Row Rue Run Shoal
		Shoals Shore Shores Skyway Spring Springs Springs Spur Spurs Square
		Square Squares Squares Station Station Stravenue Stravenue Stream
		Stream Street Street Streets Summit Summit Terrace Throughway Trace
		Track Trafficway Trail Trail Tunnel Tunnel Turnpike Turnpike Underpass
		Union Unions Valley Valleys Via Viaduct View Views Village Village 
		Villages Ville Vista Vista Walk Walks Wall Way Ways Well Wells
	)],
);

=item street_name

Return a fake street name.

=cut


__PACKAGE__->register_plugin(
	street_name	=> [
		'$last_name $street_suffix',
		'$first_name $street_suffix',
	],
);

=item street_address

Return a fake street address.

=cut

__PACKAGE__->register_plugin(
	street_address => [
		'##### $street_name',
		'##### $street_name',
		'##### $street_name',
		'##### $street_name',
		'##### $street_name Apt. ###',
		'##### $street_name Suite ###',
		'##### $street_name \####',
		'##### $secondary_unit_designator',
	],
);

=item secondary_unit_designator

Return a random secondary unit designator, with a range if needed (secondary
unit designators are things like apartment number, building number, suite,
penthouse, etc that differentiate different units with a common address.)

=cut

__PACKAGE__->register_plugin(
	secondary_unit_designator => [
		'Apartment $secondary_unit_number',
		'Building $secondary_unit_number',
		'Department $secondary_unit_number',
		'Floor #', 'Floor ##', 'Floor ###',
		'Floor \##', 'Floor \###', 'Floor \####',
		'Hangar $secondary_unit_number',
		'Lot $secondary_unit_number',
		'Pier $secondary_unit_number',
		'Room $secondary_unit_number',
		'Slip $secondary_unit_number',
		'Space $secondary_unit_number',
		'Stop $secondary_unit_number',
		'Suite $secondary_unit_number',
		'Trailer $secondary_unit_number',
		'Unit $secondary_unit_number',
		'Basement','Front','Lobby','Lower','Office',
		'Penthouse','Rear','Side','Upper',
	],
);

=item secondary_unit_number

Return a random secondary unit number, for the secondary unit designators that
take ranges.

=cut

__PACKAGE__->register_plugin(
	secondary_unit_number => [
		('A' .. 'Z'), '###','\####','##','\###','#','\##',
		'#-A','\##-A','#-B','\##-B','#-C','\##-C','#-D','\##-D','#-E','\##-E',
		'#-F','\##-F','#-G','\##-G','#-H','\##-H','#-I','\##-I','#-J','\##-J',
	],
);

=back

=head1 SEE ALSO

L<Text::FakeData>

=head1 AUTHOR

Jason Kohles, E<lt>email@jasonkohles.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2004-2005 by Jason Kohles

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

1;


lib/Text/FakeData.pm
package Text::FakeData;
our $VERSION = '0.06';

=head1 NAME

Text::FakeData - Perl extension for generating fake data

=head1 SYNOPSIS

  use Text::FakeData;

  my $faker = Text::FakeData->new();

  print "Name:    ".$faker->name."\n";
  print "Company: ".$faker->company."\n";
  print "Address: ".$faker->street_address."\n";
  print "         ".$faker->city.", ".$faker->state." ".$faker->zip."\n";

=head1 DESCRIPTION

This module creates fake (but reasonable) data that can be used for things
such as filling databases with fake information during development of
database related applications.

=cut

use strict;
use warnings;
use File::Spec ();
use Carp 'croak';
our %plugins;
my @always_import;

=head1 OBJECT METHODS

=over 4

=item new()

Object constructor.  As a shortcut, you can pass names of plugin modules to
load to new(), although this does not actually restrict the functions available
to the object, it just causes those plugins to be loaded if they haven't been
loaded already. All Text::FakeData objects in one interpreter share the plugin
data, so that multiple objects don't multiply the memory requirements.

=cut

sub new {
	my $pack = shift;
	my $self = {};
	bless($self,$pack);
	my @import = (@_,@always_import);
	unless(@import) { push(@import,'*'); }
	foreach my $import (@import) {
		foreach(@INC) {
			for(glob(File::Spec->catfile($_, qw/Text FakeData/,"$import.pm"))) {
				print "$_\n";
				require $_ if -f $_ && -r _;
			}
		}
	}
	return $self;
}

sub import { my $self = shift; push(@always_import,@_); }

=item methods();

Return a list of the methods that have been provided by all of the loaded
plugins.

=cut

sub methods { return keys %Text::FakeData::plugins; } 

=item register_plugin();

Plugin modules call register_plugin() to provide data methods.  See any of
the included plugin modules for examples.

=cut

sub register_plugin {
	my $self = shift;
	my @functions = @_;

	push(@{$Text::FakeData::plugins{shift()}}, shift()) while @_;
}

our $AUTOLOAD;
sub AUTOLOAD {
	my $self = shift;
	my $al = $AUTOLOAD;
	$al =~ s/.*:://;
	my @data = @{$Text::FakeData::plugins{$al} || []};
	croak "No data found for method '$al'" unless @data;

	my $data = $data[rand(@data)];

	my $result;
	if(! ref($data)) {
		$result = $data;
	} elsif(ref($data) eq 'ARRAY') {
		$result = $data->[rand(@{$data})];
	} elsif(ref($data) eq 'CODE') {
		$result = $data->($self);
	} else {
		croak "Don't know what to do with result of type '".ref($data)."'";
	}
	$result =~ s/\0//g;

	# replace any tokens that need expansion
	$result =~ s/\\\$/\0/g;
	while($result =~ /\$(\w+)/) {
		my $what = $1;
		my $r = $self->$what();
		$result =~ s/\$$what/$r/;
	}
	$result =~ s/\0/\$/g;

	# replace any number needing expansion
	$result =~ s/\\#/\0/g;
	$result =~ s/#/int(rand(10))/ge;
	$result =~ s/\0/#/g;
	
	return $result;
}

sub DESTROY {}

=back

=head1 LOADING PLUGINS

You can specify which plugins to load by including just the base part of their
name as an argument when loading the module with 'use'.  For example if you
only wanted to use data from the Text::FakeData::Name module, you would load
Text::FakeData like this:

  use Text::FakeData qw(Name);

By default any modules matching Text::FakeData::* in any directory in @INC
will be loaded.  You can also pass plugin names when calling the new() method,
and they will be loaded if not already in memory.  See L<new()>.

=head1 WRITING PLUGINS

Writing a plugin to provide new kinds of data is easy, all you have to do is
create a module named Text::FakeData::SomeModuleName that inherits from
Text::FakeData.

To provide data, the plugin merely needs to call the register_plugin function
with one or more pairs of function name and function data, like this:

  #!/usr/bin/perl -w
  use strict;
  use warnings;
  use Text::FakeData;

  my $faker = Text::FakeData->new();
  print "My fake data is ".$faker->some_data_function."\n";

  package Text::FakeData::SomeData;
  use base 'Text::FakeData';

  __PACKAGE__->register_plugin(
    some_data_function => [qw(foo bar baz gazonk)],
    another_data_item => sub { return '$some_data_function' },
  );

The first argument is the method that will be made available to your object,
the second is a data source.  If the data source is not a reference, it will
simply be returned as the data, if it is a reference to an array, a random
element from the array will be returned, and if it is a subroutine reference,
the subroutine will be run and the results will be returned.  The data that
your data source provides is checked for two things, tokens (that look like
perl variables, starting with a $), and numeric indicators (#).  Any tokens
found will be replaced with their values, and any numeric indicators will be
replaced with random numbers.  You can include a literal $ or # by prefacing
it with a backslash.  If you load more than one module that defines the same
function, it has an additive effect, when the function is called one of the
data sources provided will be selected at random and then it will be called
to get a piece of data.

Some data source examples:

  __PACKAGE__->register_plugin(
    age              => ['#','##'],
    monetary_amount  => ['\$####.##','\$###.##', '\$##.##', '\$#.##'],
    adult_age        => sub { int(rand(70)+18) },
  );

If your data source is a code reference, it will receive the calling object
as an argument so you can build data out of other data if you need to.  See
L<Text::FakeData::DateTime> for some examples of this.

=head1 BUGS AND KNOWN ISSUES

There is no way to selectively remove data sources from a plugin that was
loaded, even if you didn't load it.

=head1 SEE ALSO

Text::Lorem

=head1 AUTHOR

Jason Kohles, E<lt>email@jasonkohles.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2004-2005 by Jason Kohles

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut

1;
